specs/interop/superchain-weth.md

```diff
@@ -0,0 +1,161 @@
+
+| Name                     | Value                                        |
+| ------------------------ | -------------------------------------------- |
+| `SuperchainWETH` Address | `0x4200000000000000000000000000000000000024` |
+| `ETHLiquidity` Address   | `0x4200000000000000000000000000000000000025` |
+
+## SuperchainWETH
+
+### Invariants
+
+TODO
+
+### Implementation Sketch
+
+```solidity
+contract SuperchainWETH is WETH98, ISuperchainERC20 {
+  L1Block internal l1Block = L1Block(Predeploys.L1_BLOCK);
+  L2ToL2CrossDomainMessenger messenger = L2ToL2CrossDomainMessenger(Predeploys.L2_TO_L2_CROSSDOMAIN_MESSENGER);
+  ETHLiquidity liquidity = ETHLiquidity(Predeploys.ETH_LIQUIDITY);
+
+  function deposit() public payable override {
+    if (l1Block.isCustomGasToken()) revert IsCustomGasToken();
+    super.deposit();
+  }
+
+  function withdraw(uint256 wad) public override {
+    if (l1Block.isCustomGasToken()) revert IsCustomGasToken();
+    super.withdraw(wad);
+  }
+
+  function sendETH(uint256 chainId) external payable {
+    deposit();
+    _sendWETH(msg.sender, msg.value, chainId, true);
+  }
+
+  function sendETHTo(address to, uint256 chainId) external payable {
+    deposit();
+    _sendWETH(to, msg.value, chainId, true);
+  }
+
+  function sendERC20(uint256 wad, uint256 chainId) external {
+    _sendWETH(msg.sender, wad, chainId, true);
+  }
+
+  function sendERC20To(address to, uint256 wad, uint256 chainId) external {
+    _sendWETH(to, wad, chainId, true);
+  }
+
+  function finalizeSendWETH(address to, uint256 wad, bool unwrap) external {
+    if (msg.sender != address(messenger)) revert Unauthorized();
+    if (messenger.crossDomainMessageSender() != address(this)) revert Unauthorized();
+
+    // Mint the required liquidity.
+    if (l1Block.isCustomGasToken()) {
+      liquidity.mint(wad);
+    }
+
+    // Mint the WETH balance.
+    balanceOf[to] += wad;
+    emit Bridge(to, wad)
+
+    // If user wants to unwrap, try unwrapping.
+    if (unwrap) {
+      try this.unwrap(to, wad) {
+        // Great.
+      } catch {
+        // Just leave it wrapped.
+      }
+    }
+  }
+
+  function unwrap(address to, uint256 wad) public {
+    if (msg.sender != address(this)) revert Unauthorized();
+
+    // Logic taken from withdraw(...)
+    require(balanceOf[to] >= wad);
+    balanceOf[to] -= wad;
+    payable(to).transfer(wad);
+    emit Withdrawal(to, wad);
+  }
+
+  function _sendWETH(address to, uint256 wad, uint256 chainId, bool wrapped) internal {
+    _burn(msg.sender, amount);
+    messenger.sendMessage({
+      _destination: chainId,
+      _target: address(this),
+      _message: abi.encodeCall(this.finalizeSendWETH, (to, amount, wrapped))
+    });
+  }
+}
+```
+
+## ETHLiquidity
+
+### Invariants
+
+TODO
+
+### Implementation Sketch
+
+```solidity
+contract ETHLiquidity {
+  address internal weth = Predeploys.SUPERCHAIN_WETH;
+  L1Block internal l1Block = L1Block(Predeploys.L1_BLOCK);
+
+  function burn() external payable {
```

Comment:
This function is not used, it existed to try to make `totalSupply` of weth more accurate

Reply:
Honestly at that point it might just be easier to give `SuperchainWETH` the massive ETH supply and call it a day. Is there a specific class of issues that we're avoiding with `ETHLiquidity`?

Reply:
I definitely prefer to have 2 contracts here but am open to changing. My reasoning is as follows:
- The design [here](https://oplabs.notion.site/Public-Interoperable-ETH-Prototype-3fb07a8310304eeb83ba217f42932eac?pvs=4) already shows it works
- Preserve ~accurate `totalSupply` which is nice for data team
- Prevent unknown unknown bugs with very explicit conditions for being able to mint
- Generally things should do 1 thing well and be composed together
- Enable `ETHLiquidity` to be a primitive that can be used by the system in other ways in the future, this is about flexibility with future upgrades